### ISSUE TITLE
Feature Request | can OIDC be added to the provider

### ISSUE THREAD

**Issue by scott-doyland-burrows** [cloudsmith-io/terraform-provider-cloudsmith#128]
Would it be possible to get OIDC on the roadmap for provider authentication.

We'd like to authenticate from Terraform Cloud to Cloudsmith using OIDC.

It doesn't look like the provider supports OIDC at all?

I know Cloudsmith can use OIDC, but we have only just started to use Cloudsmith so I am not sure if this would need just a provider change, or also a change to Cloudsmith?

Something similar to this please:
https://developer.hashicorp.com/terraform/cloud-docs/workspaces/dynamic-provider-credentials/aws-configuration

**Comment by cloudsmith-iduffy** [comment]
Hi Scott,

Great news! What you're looking to achieve is already possible by combining Terraform Cloud's [workload identity token generation](https://developer.hashicorp.com/terraform/cloud-docs/workspaces/dynamic-provider-credentials/manual-generation) with Cloudsmith's [OpenID Connect](https://help.cloudsmith.io/docs/openid-connect) functionality.

Let me walk you through the implementation I've created. I've set up a complete working example in my repository at [terraform-cloud-oidc](https://github.com/cloudsmith-iduffy/terraform-cloud-oidc), which demonstrates how to exchange a Terraform Cloud JWT for a Cloudsmith JWT using a data provider.

The setup process involves two main components:

First, we need to configure the OIDC provider in Cloudsmith. Based on the JWT specifications detailed in [Terraform Cloud's documentation](https://developer.hashicorp.com/terraform/cloud-docs/workspaces/dynamic-provider-credentials/workload-identity-tokens), I've configured the Cloudsmith OIDC provider with these settings:

```json
Provider URL: https://app.terraform.io
Claims Configuration:
{
    "aud": "cloudsmith",
    "terraform_organization_name": "iduffy-demo"
}
```

<img width="647" alt="Screenshot 2024-12-01 at 16 58 05" src="https://github.com/user-attachments/assets/f3ccc3ff-c6be-4077-b838-0a77d9cb4864">

Second, we need to enable the workload identity feature in Terraform Cloud. This is done by setting an environment variable named `TFC_WORKLOAD_IDENTITY_AUDIENCE` with the value `cloudsmith`. 

Once these components are in place, the system works seamlessly: when your Terraform code executes in Terraform Cloud, it automatically uses the data resource to obtain a Cloudsmith token, which is then passed to the Cloudsmith provider. You can find more details about this process in the [Manually generating workload identity tokens](https://developer.hashicorp.com/terraform/cloud-docs/workspaces/dynamic-provider-credentials/manual-generation) documentation.

<img width="1131" alt="Screenshot 2024-12-01 at 16 58 48" src="https://github.com/user-attachments/assets/fd00a805-5e16-40a3-8536-785ce5dd3bee">

Please let me know if you'd like me to clarify any part of this setup or if you have questions about implementing it in your environment.

<img width="1152" alt="Screenshot 2024-12-01 at 16 59 38" src="https://github.com/user-attachments/assets/151d99c7-129a-4382-bdb8-6aae3b4ec21f">

**Comment by scott-doyland-burrows** [comment]
Thanks.  That worked.

It would be nice if this could work without the need for using the bash script though.

Our organization is looking to prevent local-exec and the ```external``` data resource from running when we use terraform cloud, as there are security risks with this, ie it is possible to display sensitive variables in clear text.

**Comment by cloudsmith-iduffy** [comment]
>It would be nice if this could work without the need for using the bash script though.

Agreed! Will discuss this with the team and see what we can come up with. Copying something along the lines of the `AWS_WEB_IDENTITY_TOKEN_FILE` used in the AWS credentials provider chain would be nice.

> Our organization is looking to prevent local-exec and the external data resource from running when we use terraform cloud, as there are security risks with this, ie it is possible to display sensitive variables in clear text.

That is completely fair. The tokens are short lived and ideally any terraform state is stored encrypted at rest but totally understand that potential exposure - sadly data resources to my knowledge can't be marked as sensitive. 

Another alternative is to use hooks to fetch and place the temporary credential into a TF_VAR environment variable https://developer.hashicorp.com/terraform/cloud-docs/agents/hooks - sadly this looks to require terraform cloud enterprise.

**Comment by scott-doyland-burrows** [comment]
Just to clarify the reason there is concern re local-exec and the external data resource. 

It makes it possible for someone with write access to a GitHub repo, to create a feature branch, put a local-exec (or external data resource) in the code, then create a GitHub tag and cause that code to be executed (if using a VCS workspace running on tag creation). 

If they run a "printenv" from the local-exec, it will print out all envars including sensitive variables. 

So, the concern isn't a short lived cloudsmith token, it is really around tokens that are static until we rotate them, such as cloudflare tokens. 

If we give someone access to the repo and access to review plans and applies, then they can access sensitive variables.

**Comment by BartoszBlizniak** [comment]
Hey @scott-doyland-burrows  — following up on this request, native HCP Terraform OIDC authentication is now available in [v0.0.84](https://github.com/cloudsmith-io/terraform-provider-cloudsmith/releases/tag/v0.0.84).

This brings the token exchange demonstrated in the earlier workaround directly into the Cloudsmith Terraform provider. The provider now reads the workload identity token injected by HCP Terraform, exchanges it for a short-lived Cloudsmith credential, and refreshes that credential when required.

As a result, you no longer need to use a Bash script, `local-exec`, or the `external` data source. The HCP Terraform workload identity token is also not passed as a provider argument or stored in Terraform state.

The provider can be configured as follows:

```hcl
provider "cloudsmith" {
  oidc {
    organization = "your-organization"
    service_slug = "your-service-account"
  }
}
```

In your HCP Terraform workspace, set `TFC_WORKLOAD_IDENTITY_AUDIENCE` (or the tagged `TFC_WORKLOAD_IDENTITY_AUDIENCE_CLOUDSMITH`) as an environment variable.

On the Cloudsmith side, configure an OIDC trust for `https://app.terraform.io`, allow the service account specified above, and configure the expected audience and suitably restrictive HCP Terraform claims, such as the organization and workspace IDs.

One important point is that the OIDC trust must exist before the provider can authenticate using it. You can either create the trust through the Cloudsmith UI, or apply the `cloudsmith_oidc` resource once using a static API key before migrating the provider configuration to OIDC.

Once you've confirmed that OIDC authentication is working, you can remove `CLOUDSMITH_API_KEY` from the workspace so that the long-lived credential is no longer required.

The full setup is documented in the provider's [OIDC authentication guide](https://registry.terraform.io/providers/cloudsmith-io/cloudsmith/latest/docs#authenticate-with-oidc).

Please let us know if you run into any issues while migrating. Otherwise, I think we can consider this delivered and close the issue.
